### ISSUE TITLE
Suggestion: `uv bundle` or similar to create a contained executable a la pyinstaller, py2exe

### ISSUE THREAD

**Issue by matterhorn103** [astral-sh/uv#5802]
I regularly use pyinstaller for a project, and the frequency of questions about it - or the task more generally of "compiling" an executable for distribution - around the web indicates to me that it's widely popular.

It would be incredibly cool if uv offered a `uv bundle` command (or whatever name) that essentially did what pyinstaller does and bundles together everything to afford an executable that contains the dependencies and interpreter and can be distributed to users as-is.

I assume that pyinstaller, py2exe etc. are complex projects and implementing such functionality would be not trivial in the slightest, so I would have thought it would be a case of using `pyinstaller` itself in the background like the way that build backends are used.

Presumably there will one day be a `uv build` command but I assume that will be similar to `rye build` in that it will create an sdist or wheel. Maybe there could be a `uv build --executable` option?

**Comment by matterhorn103** [comment]
Some relevant links:

- It seems someone [asked for the same in Rye](https://github.com/astral-sh/rye/issues/325) previously
- ofek's [PyApp](https://github.com/ofek/pyapp) and indy-greg's now-dead [PyOxidizer](https://github.com/indygreg/PyOxidizer) are other, more modern executable generators, both seem simpler than pyinstaller, and both are written in Rust.
- Hatch seems to have implemented essentially this, as part of a [Hatch plugin](https://hatch.pypa.io/latest/plugins/builder/binary/) that uses PyApp, and it is invoked there by having "binary" as a build target

**Comment by chrisrodrigue** [comment]
To add… despite its popularity, `pyinstaller` suffers from slowness because has to unzip all the dependencies and the interpreter when the executable is launched. It offers users the option to display a splash image to distract them from the perceived slowness.

**Comment by matterhorn103** [comment]
@my1e5 why do you say that Nuitka would be unsuitable for inclusion with uv?

**Comment by my1e5** [comment]
@matterhorn103, I guess what I mean is that Nuitka is perhaps quite an opinionated way of packaging your Python code into an executable - as it is essentially an alternate Python interpreter which compiles Python to C. Whereas tools like Pyinstaller are more like a bundler - taking the Python interpreter, Python files and dependencies and packaging them into an executable. This is definitely more straightforward than the Nuitka approach, but it does make the Pyinstaller executable very easy to un-package and see the underlying source code (see https://github.com/extremecoders-re/pyinstxtractor). Which might be a caveat needed when including certain 'executable creators' within uv - you need to make users aware that their exe can easily be un-packaged.

In terms of licensing, Nuitka is MIT licensed. But it does also have a commercial tier - which may or may not complicate matters if Astral were to want to bundle it with uv, I don't know.

**Comment by zanieb** [comment]
I think this is a duplicate of https://github.com/astral-sh/uv/issues/2799, though there's more discussion here.

**Comment by martimlobao** [comment]
Also want to call out pex, a tool for generating .pex files (Python EXecutable) which are self-contained zipped executable Python environments containing sources, requirements, and dependencies.

https://github.com/pex-tool/pex

**Comment by jbvsmo** [comment]
I am interested in a solution like this. I think the bundler should be somewhat coupled with `uv run` since you want to turn any entry point into an executable. 

Example
```bash
$ uv init --script foo.py
$ uv add requests --script foo.py
$ uv run foo.py
Hello from foo.py!
```

Then you could make an executable which is composed of
1. bootstrap code to download uv if needed and run it
2. target code 
3. uv executable **(optional)**
4. python executable **(optional from `uv python`)** 
5. populated venv **(optional)**

```bash
$ uv bundle foo.py -o foo --include-uv --include-python --include-dependencies
$ ./foo 
Hello from foo.py!
```
This above would be a self contained binary (roughly 30MB as uv is ~12MB and python ~17MB).

If downloading stuff on the fly is ok, the bootstrap code would just download uv, then uv would download python, install dependencies and run all during 1st execution.

It should also be possible to create binaries cross platform since uv supports install/sync with `--python-platform`

```bash
$ uv bundle foo.py -o foo.exe --python-platform windows
```

_In a 2nd iteration_, it would be nice to be have installers such as to create .dmg (mac) or an .exe windows installer file.

**Comment by dbrtly** [comment]
what if the command was `uv run ruff --output-file ruff.uvx`
with the -o/--output agument as an indicator to create the binary file rather than execute this tool right now.

**Comment by stefanondisponibile** [comment]
I love this idea and @jbvsmo proposal for CLI usage.

I am a bit confused on the best choice for the bundling "backend" though, I see different options on the table and I'm not sure, how would you choose one over the other? Should maybe `uv` propose some protocol for each library to support bundling via `uv`?

This feature would be great also for bundling pyspark apps (thinking about pex support mentioned by @martimlobao here)

**Comment by AKuederle** [comment]
I think the bootstrap option mentioned by @jbvsmo is almost achievable with the existing uv options.

You could do the following:

Within your project folder have a script (written in the scrippting language supported by your operating system) that does the following:

1. Unmanaged installation of uv (https://docs.astral.sh/uv/configuration/installer/#unmanaged-installations) can be used to just get the uv executable in a subdir of the project folder.
2. Setup the cache dir to be inside your project folder (https://docs.astral.sh/uv/concepts/cache/#cache-directory)
3. Then install Python
4. Install project deps
5. Use uv run to execute a script within the project enviroment. The easiest option would be to just "run" the package using a `__main__.py` within the project

Step 1 could of course be skipped, if the uv executable is already recognized within the specified subdir. And step 3-5 are basically what happens when you run uvx. So maybe this is actually a 2 line script.

The great thing is that this script would be project independent! So someone just would need to write a script for each operating system and you could throw these scripts in your project folder.

Then you just need to distribute this project folder and have people execute the script corresponding to their operating system.

This process could then of course be wrapped further by an installer, but seems relatively straight forward to at least address the cases of "Please provide an exe"

**Comment by jromal** [comment]
For me it would be enough to have `uv run` with some options.

For example I have done a program (as module) using PySide6.
I can do `uv run -m my_module` from the folder above the module and it works. Fast, efficient, impressive.
But there are two things that can help as an intermediate solution before a `uv bundle`.
* Read a compressed file for the full project: `uv run --source my_file.zip`
* Allow an option to hide or close the terminal for qt programs: `uv run --source my_file.zip --no-terminal`

With this solution I could distribute on Windows the files with "Inno Setup", adding `uv.exe`, the zip file, and maybe a ".bat" file with a logo to execute the `uv -run` command.

**Comment by johnthagen** [comment]
As much as I like Pyinstaller (to me it's best of current options), one other thing that it suffers from is that anti-virus vendors are _always_ flagging generic Pyinstaller binaries as malware. This is (at least in part, I think) because Pyinstaller uses the same bootloader binary for all built applications, thus if someone bundles malware with Pyinstaller version X, all other users who bundle legitimate Pyinstaller applications can get flagged as malware too. I've seen this hit AV software across both Windows and Linux over the years. Not sure I have any ideas on how Astral could improve _that_ issue, but wanted to bring it up as a current (very annoying) real world challenge.

**Comment by ceejatec** [comment]
It's not exactly the same thing, but I've suggested a relatively simple feature to `uv` that would solve at least a subset of the requirements of delivering a single downloadable executable for a Python program: #10465

Basically it would let `uv` serve as a limited runtime-only alternative to PyApp.

**Comment by jromal** [comment]
Thanks for your request and for commenting here.

For me the final tool would be a third party tool, part of uv (for Windows, uv.exe, uvx.exe and the new one, let's call it `uvd.exe`).

And that third tool should be `PyOxidizer`. I do not know if you know it. But the single python binaries that Astral uses are a "byproduct" of PyOxidizer. A project from Gregory Szorc (indygreg on GitHub). It is a rust tool for byilding executables including a full python run time environment. Gregory had no time for any of the projects. The python binaries are now generated by Astral. PyOxidizer is not maintained since many months. I considered it as a future better alternative to Pyinstaller or Nuitka or ...

Maybe we could get a simple tool and a complex one with all the hooks for the different packages. Maybe could also have different license models to allow Astral to capitalize the investment. But should be continued as a natural evolution of uv.

But as a you say a wrapper of uvx would do it.

Let's wait and see.

**Comment by Nardol** [comment]
Unfortunately https://github.com/indygreg/PyOxidizer only support Python up to 3.10 as specified in the latest stable documentation, except that it looks promising. Latest commit is two months old but you say it is not maintained again?

**Comment by paul-gauthier** [comment]
I would also be interested in more direct support for using uv as an "installer" for python CLI tools.

I've been using uv in novel ways for this purpose already. I recently wrote a bit about these approaches. Leaving the link here in case folks find it interesting or helpful:

https://aider.chat/2025/01/15/uv.html

**Comment by thewh1teagle** [comment]
It will be useful if uv add `compile` command like in bun.js which will produce self contained executable.
The only important thing is NOT to work like pyinstaller extract the contents but somehow load the interpreter into the memory, and tell cpython to run the bundled script directly from the memory.
Just like bun.js / nodejs does.

**Comment by davidlanouette** [comment]
It's worth noting that it looks like PyOxidizer has been taken over ty Astral-sh - in the form of [python-build-standalone](https://github.com/astral-sh/python-build-standalone).

So, this issue can likely be closed.

**Comment by zanieb** [comment]
No, we don't own PyOxidizer — we're just maintaining the standalone Python distributions. These ideas are separate.

**Comment by davidlanouette** [comment]
Interesting.  That's definitely not how I interpreted the [blog post](https://astral.sh/blog/python-build-standalone)

> TL;DR: On 2024-12-17, we'll be taking stewardship of [python-build-standalone](https://github.com/indygreg/python-build-standalone), [Gregory Szorc's](https://gregoryszorc.com/) foundational project for building and installing portable Python distributions. You can read Gregory's own announcement [here](https://gregoryszorc.com/blog/2024/12/03/transferring-python-build-standalone-stewardship-to-astral/).

But, it does explain why the releases seem to only contain python distributions.

**Comment by jdegenstein** [comment]
This is a bit off topic, but I had a related discussion on discord about this topic and thought I would share A solution to MY problem that could be relevant to others. I feel that `python-build-standalone` has huge potential QoL improvements for the python package management ecosystem because it (finally) seems to attempt to address the local vs. global installation "problem".

My problem statement: create a zipped folder with standalone python, standalone packages and their dependencies that I can distribute to other users ("download, unzip and run"). Non-goals: a single executable -- for me a folder containing lots of subdirectories/files is fine but an obvious "entry point" is desirable. Obviously the result will be OS and CPU type specific.

Steps I followed:
1. Download relevant https://github.com/astral-sh/python-build-standalone/releases and unzip
2. chdir to python subdir
3. execute e.g. `python.exe -m pip install uv` (there is no specific need to use uv by the way)
4. execute e.g. `python.exe -m uv pip install <somepackage>`
5. for my application I created `somepackage.bat` (one directory "up") containing `call python/python.exe -m <somepackage>`. Obviously shell scripts would be useful on other platforms, and I am sure there are other solutions as well.
6. zip the parent folder and distribute to others with same OS and CPU type

The reason I like this solution is multifold (1) uses standard package installation methods (2) everything (?) is local to the folder (3) no e.g. pyinstaller-hooks-contrib to maintain. Are there downsides to this approach? Probably, but in my quick testing did not run across any issues -- it "just worked". This is contrast to my prior experiences with nuitka and pyinstaller which work perfectly for projects with pure python dependencies, but can require more "handholding" when that is not the case.

Finally back to the topic of this post -- this would be very nice if supported by uv directly in some form.

p.s. thanks to @geofft for their help

**Comment by jromal** [comment]
The creator of the standalone python (Gregor Szorc) did it for his tool PyOxidizer. A rust substitute of pyinstaller or Nuitka. 

I am repeating when I can that Astral should take it and finish it. Gregor (indygreg) has no time and his work was really good. Just unfinished. But he did one of the biggest changes in python. The standalone python. Do not let his work disappear.

**Comment by dbrtly** [comment]
@jdegenstein it sounds like you want to do `uv tool install my_app` but have it work even if uv is not installed on the client machine.

**Comment by charliermarsh** [comment]
(@jromal -- incase you haven't seen, we did assume stewardship of python-build-standalone: https://github.com/astral-sh/python-build-standalone)

**Comment by tjnd89** [comment]
@matterhorn103 FWIW, this is exactly my use case. I'm building a an "app" for non-programmers. I use uv with hatch to build the distributions. I have a custom hatch build script that kicks off pyinstaller to build the "executable". In my case, I'm using the "one_dir" option because along with the exectuable I have to ship various data files that are needed. One thing I ran into is that running pyinstaller standalone from the command line worked great. My custom build script that issued the same command line arguments was failing to find all the dependencies because of uv's build isolation. I had to set 

[tool.uv]
no-build-isolation = true

and then it all worked well. Here's the snippets of relevant code.

hatch_build.py

```
import subprocess

from hatchling.builders.hooks.plugin.interface import BuildHookInterface


class CustomHook(BuildHookInterface):
    """USed to kick off Pyinstaller as part of a standard uv build.
    """
    def initialize(self, version, build_data):
        cmd_options = ['pyinstaller',
                       '--noconfirm',
                       # add your command line options here
                       'blah.py'   # your script you are wrapping.
                       ]

        subprocess.run(cmd_options, check=False)
```

pyproject.toml

```
[tool.hatch.build.targets.sdist.hooks.custom]
    dependencies = ["pyinstaller"]
```

This should get you what want I think.

**Comment by jdegenstein** [comment]
@dbrtly Yes, that is correct -- in my experience with developing apps for new python users or python beginners as soon as the phrase "virtual environment" is uttered about half of the potential users will just give up (again my stated goal for my users is "download (my app), unzip and run"). Single executables are nice, as are installation wizards, but that introduces another layer of OS specific considerations that are not worth it for me (yet).

**Comment by dbrtly** [comment]
It seems that you gave to draw the line somewhere and make an assumption about pre-requisites on the target machine: git? Is user local admin? Is the user sophisticated enough to use a cli tool at all? Is pyinstaller available? Is uv available?

You could go a long way with something like this:

```
# install uv and put it on path
uv tool install my_app \
  --default-index https://custom.pkg.dev/index 
```

Am curious about what specific features you can see potential for uv to improve on for this user experience.

**Comment by chrisrodrigue** [comment]
I posted some ideas over in #11746 about what I think a `uv layout` command might look like.

`uv bundle` would be distinct from `uv layout` and both would be valuable to developers.

From [my comment](https://github.com/astral-sh/uv/issues/11746#issuecomment-2681843706):
> I think the intent of `uv bundle` is to be able to combine an app, its dependencies, and Python for non-developer end users. It would produce an executable binary, likely having a GUI or CLI. This would enable devs to distribute their apps without relying on users to know what uv, Python, or PyPY is.
>
>The intent of this proposal ([maybe uv layout for parity with other tools](https://learn.microsoft.com/en-us/visualstudio/install/create-an-offline-installation-of-visual-studio?view=vs-2022#step-2---create-a-local-layout)) is to be able to combine uv, packages, and Python for developers. It would produce a directory, likely having a uv.toml to configure uv to work with the directory layout when invoked. This would enable devs to create projects without relying on an internet connection.
>
>Both of these commands would be valuable to developers, with `uv bundle` artifacts targeted at users and `uv layout` artifacts targeted at developers. Both commands could read a `pyproject.toml` to produce the desired artifact for the current platform by default, or take any number of `--platform <target-triple>` options (or `--platform all`) to produce artifacts for different platforms.

`uv compile` could also be good idea, distinct from both `uv bundle` and `uv layout`. It would be similar to `uv bundle`, but it might produce an artifact more akin to that of `nuitka` rather than `pyinstaller` and would be much more complicated to generalize and implement.

**Comment by chrisrodrigue** [comment]
Summary of some of the ideas mentioned. The command names are flexible so I'm just using these for now to tell them apart 🙂

| Command | Use-case | Behavior | Similar to | Implementation(s) | 
| --- | --- | --- | --- | --- | 
| `uv bundle` | To create a portable executable for users | Outputs a platform-specific or actually portable executable ([APE](https://justine.lol/ape.html)) that embeds the interpreter and dependencies | [pyinstaller](https://pyinstaller.org/en/stable/), [pyapp](https://github.com/ofek/pyapp), [pyoxy](https://github.com/indygreg/PyOxidizer) | On-disk execution (slower?), [in-memory execution](https://en.wikipedia.org/wiki/In-memory_processing) (faster?) | 
| `uv compile` | To create a portable executable for users | Outputs a platform-specific or actually portable executable ([APE](https://justine.lol/ape.html)) without embeddings | [Nuitka](https://nuitka.net/) | Compilation, transpilation | 
| `uv layout` | To create a portable Python distribution for developers | Outputs a directory that contains platform-specific uv/interpreter executables and Python packages | [WinPython](https://github.com/winpython/winpython), [Anaconda](https://docs.anaconda.com/anaconda/release-notes/), [ActiveState](https://www.activestate.com/platform/supported-languages/python/) | `fs::create_dir` |

**Comment by paveldikov** [comment]
@chrisrodrigue `uv layout` can be approximated by stitching together `uv venv --relocatable` with `uv python download` + a launcher script that rewrites `pyvenv.cfg` (because the base interpreter definition is not portable -- you can specify relative paths but those are implicitly relative to the current working directory)

I have done this for a proof-of-concept and, whilst clunky, it definitely does work. In fact, by repackaging its output into something like an RPM or a self-extracting executable, you end up with a passable (though non-optimised) `uv bundle` too.

**Comment by chrisrodrigue** [comment]
> [@chrisrodrigue](https://github.com/chrisrodrigue) `uv layout` can be approximated by stitching together `uv venv --relocatable` with `uv python download` + a launcher script that rewrites `pyvenv.cfg` (because the base interpreter definition is not portable -- you can specify relative paths but those are implicitly relative to the current working directory)
> 
> I have done this for a proof-of-concept and, whilst clunky, it definitely does work. In fact, by repackaging its output into something like an RPM/DEB/Choco/pick-your-favourite-OS-package-manager -- or a self-extracting executable -- you end up with a passable (though non-optimised) `uv bundle` too.

Interesting, thanks @paveldikov. `uv python download` doesn’t appear in the [docs](https://docs.astral.sh/uv/reference/cli/#uv-python), but I’ll try experimenting with it today. 

An open question I have about venvs… how can install packages from one venv into another? AFAIK, there is no way to do this because you need wheels to do real installation into a venv. 

You could add the source environment to PATH and make new venvs with `--system-site-packages` to use packages from the source environment, but there are downsides:
- `--system-site-packages` is an all or nothing approach. You get all packages regardless of whether you need all of them.
- System site packages aren’t installed to the venv, so you can’t pack up your project to share.
- All bets are off with the tooling. You need to manually check that any dependencies in `pyproject.toml` are in the source environment you’re using, which adds toil.

These reasons are why I advocate for using wheels with layouts. However, venvs may be better for bundles since they compress better. The checksum of the bundle executable should be good enough for attestation.

**Comment by chrisrodrigue** [comment]
I tried running `uv pip download` and `uv python download` to manually create a layout that can be bundled/packed/archived, but alas, neither of these commands exist.

First, you will want `uv`:
```
curl -LSsf uv.zip https://github.com/astral-sh/uv/releases/download/0.6.4/uv-x86_64-pc-windows-msvc.zip | tar -x
```

To grab wheels in lieu of `uv pip download`: 
```
`python -m pip download -r requirements.txt -d pypackages` 
```

T o grab the interpreter in lieu of `uv python download`:
```
curl -LSsf https://github.com/astral-sh/python-build-standalone/releases/download/20250212/cpython-3.13.2+20250212-x86_64-pc-windows-msvc-install_only_stripped.tar.gz | tar -x
```

Configuring `uv` use the `python` and `pypackages` is the real bounty 😁 I feel like a `uv.toml` adjacent to the executable would be appropriate for configuration.

**Comment by paveldikov** [comment]
Sorry -- it's `uv python install`... some poetic licence is necessary when citing commands off by heart 🙃

It will give you a known-good relocatable python site.  You can then manipulate that as you please, e.g. by further relocating it to sit inside of a relocatable venv. And then manipulate `pyvenv.cfg` accordingly.

**Comment by chrismatix** [comment]
>  @martimlobao Also want to call out pex, a tool for generating .pex files (Python EXecutable) which are self-contained zipped executable Python environments containing sources, requirements, and dependencies.
> https://github.com/pex-tool/pex

To anyone who was looking into this option: I put together a recipe repository for bundling uv workspaces by compiling their dependencies using `uv pip compile` and then using that lock file to build an executable with `pex`: https://github.com/chrismatix/uv-pex-monorepo

If you check it out, please let me know about any edge cases it does not accommodate!

**Comment by luigi311** [comment]
Great ideas guys. I went with something similar on my project. `uv build` which generates a whl file i then used pex on that whl and then finally used fpm to generate a deb for me to distribute. Hopefully something built into uv comes that makes this more simple but for now this works without complicating my source files with duplicate files that need to be manged. Can find my initial implementation here https://github.com/luigi311/immich_upload_daemon/blob/d1ec12c369ab4192e0ba88653010e4a0968e6e26/.github/workflows/build.yml

**Comment by TanixLu** [comment]
@AKuederle My thoughts align closely with yours: create a bootstrap-style APE that downloads `uv`, installs Python, fetches dependencies, and executes the source code. I've already implemented this in a package called [pyfuze](https://github.com/TanixLu/pyfuze).

**Comment by blais** [comment]
> [@AKuederle](https://github.com/AKuederle) My thoughts align closely with yours: create a bootstrap-style APE that downloads `uv`, installs Python, fetches dependencies, and executes the source code. I've already implemented this in a package called [pyfuze](https://github.com/TanixLu/pyfuze).

I don't think this is what's needed; this requires internet access and (correct me if I'm wrong) doesn't build binaries.
What's needed:
- pre-build all the binaries
- compute the transitive closure of all desired packages
- include them all in some archive format
- make it possible to import from the archive format
- create a stub at the front of it to make it runnable
The idea is that the package should be runnable without a python install and without having to install stuff from the internet (so it can be run at scale on the cloud / in a VPC)

**Comment by TanixLu** [comment]
> * The idea is that the package should be runnable without a python install and without having to install stuff from the internet (so it can be run at scale on the cloud / in a VPC)

This is a trade-off. If everything is prepared in advance instead of being downloaded online, the compressed package will be large in size, and cross-platform compatibility will become problematic. This undermines the advantages of APE's cross-platform capability and uv's fast dependency downloading. For users without programming experience, having a single file that runs on all platforms is incredibly convenient. If pre-prepared cross-platform uv, Python, and dependencies are required, and APE selects them at runtime based on the operating system, the package would be even larger.  

With the current version of [pyfuze](https://github.com/TanixLu/pyfuze), you could run it once on a specific platform and then recompress it into a package, eliminating the need for an internet connection. However, the trade-off is that the application becomes platform-specific and larger in size. Perhaps I should add appropriate options to pyfuze, allowing users to opt to pre-download dependencies.  

Additionally, I noticed that [APE-is-also-a-ZIP](https://github.com/jart/cosmopolitan/issues/166). I wonder if it could be made self-extracting, which would be even more convenient.

**Comment by blais** [comment]
No. The trade-off defines a completely different feature ask. I think you have an uncommon definition for the meaning of "bundling."

There's a lot of prior art on this. See Bazel "par" files and XAR files and a bunch of other packagers for Python. The ability to run an _isolated_ package without any network access is the key feature of "bundling," the isolation is the feature that allows reliable execution on a variety of hosts. Especially for cron jobs. In a private cluster with tight security you cannot rely on uv fetching resources available on the internet, resources that can change between invocations, or even uv's own behavior to change. I would argue that even if you pinned down the version of uv itself in the zip file you're possbly liable to see failures. Bundles are basically single-application lightweight docker containers, where you don't have docker or you don't want its limitations (e.g., mount/network isolation) -- you just want to run the program from a single file as if it were an a single executable file, with no dependencies.

The bundling that you provide in pyfuze is minimal (I just read your code). It's a wrapper that copies the script and a few files configuring the dependencies for uv to do the heavy lifting. I think its usefulness is limited; call it "lightweight bundling" perhaps. Except for the uv install bit (which really ought to be done once per host), a regular Python program with a main file that has a `#!/usr/bin/env -S uv run --script` hash-bang does most of what pyfuze provides IMHO.

These are two separate sets of problems. "Bundling" is the former.

**Comment by kskarthik** [comment]
This feature would be a great addition to the python ecosystem! 

Deno & Bun implemented it for the node ecosystem.

**Comment by parkan** [comment]
being able to ditch all the glue code needed by pyinstaller would be *amazing*

**Comment by kishaningithub** [comment]
Pyapp can also be considered an option for creating statically linked binaries

https://www.infoworld.com/article/4030697/pyapp-an-easy-way-to-package-python-apps-as-executables.html

**Comment by parkan** [comment]
> Pyapp can also be considered an option for creating statically linked binaries
> 
> https://www.infoworld.com/article/4030697/pyapp-an-easy-way-to-package-python-apps-as-executables.html

this is looks great and appears to natively (though optionally) use `uv` in the build chain https://ofek.dev/pyapp/latest/runtime/

will definitely try this out!

**Comment by alekssamos** [comment]
I made an sfx file using winrar. Set the options to show nothing, hide everything, and accept the defaults. Execute after installation `uv run app.py`. And it worked.

**Comment by metaist** [comment]
Took me a while to rewrite the whole thing to use `uv`, but if you add an appropriate `[project.scripts]` to your `pyproject.toml` then you can do:

uvx cosmofy bundle

And it'll produce a single Cosmopolitan Python for each entry point using uv. You can also do `--script` to bundle up a single script with its inline dependencies. There are [some limitations](https://github.com/metaist/cosmofy#limitations), but it's a start. 

See: https://github.com/metaist/cosmofy

**Comment by inoa-jboliveira** [comment]
Hi @metaist I'm interested in your solution, and might give it a try. One thing I noticed is that you seem to be [hosting a python binary yourself](https://github.com/metaist/cosmofy/blob/815ee899788b3e062a014960c35901b9181d1ea1/src/cosmofy/bundle.py#L38) instead of using uv to download it which would manage correctly the arch and OS.

Having this is also not auditable and some places would not accept such file. Is there any attestation where it comes from?

**Comment by metaist** [comment]
> Hi [@metaist](https://github.com/metaist) I'm interested in your solution, and might give it a try. One thing I noticed is that you seem to be [hosting a python binary yourself](https://github.com/metaist/cosmofy/blob/815ee899788b3e062a014960c35901b9181d1ea1/src/cosmofy/bundle.py#L38) instead of using uv to download it which would manage correctly the arch and OS.
> 
> Having this is also not auditable and some places would not accept such file. Is there any attestation where it comes from?

Thanks @inoa-jboliveira!

1. The whole point of [**Cosmopolitan** Python](https://ahgamut.github.io/2021/07/13/ape-python/) is that it doesn't depend on arch and OS. It's a single binary file that is actually a very clever zip that runs on all the OSes without changes. Part of the reason I wanted to build this is so that I don't have to rebuild different binaries for every OS. It's just one binary. Note that [I do use `uv` to build a temporary venv](https://github.com/metaist/cosmofy/blob/815ee899788b3e062a014960c35901b9181d1ea1/src/cosmofy/bundle.py#L227) of your project (or script) which captures all the dependencies properly on the OS you happen to be on and then I bundle them into the Cosmopolitan Python executable.

2. The point about audits/attestation is valid and strong; this is one of the limitations of the current approach (see https://github.com/metaist/cosmofy/issues/44). I'm currently using the python binary built and hosted by Justine Tunney (@jart) through [superconfigure](https://github.com/ahgamut/superconfigure) by @ahgamut. Every single version of python has to be rebuilt to be a Cosmopolitan Python (superconfigure has lots of patch files to convert the paths and make other adjustments) and right now they're only up to 3.12.3. A longer-term goal I have is to make a dedicated repo (superconfigure builds lots of binaries besides python) for tracking all the changes and having a single attestable build per python version similar to how uv used and then took over [`python-build-standalone`](https://github.com/astral-sh/python-build-standalone).

3. This is also why you can override the default Cosmopolitan Python URL through `COSMOFY_PYTHON_URL`  or `--python-url` to point at a Cosmopolitan Python binary that you've verified or self-host.

**Comment by inoa-jboliveira** [comment]
>      [**Cosmopolitan** Python](https://ahgamut.github.io/2021/07/13/ape-python/)

I had no idea what this is and still I'm kinda intrigued (being nerdsniped here). At the same time, this is not what I'm looking for right now as it will not support C/Rust compiled libraries.

**Comment by metaist** [comment]
Yes, it's kinda wild that with the right combinations of magic numbers this is possible. 

C extensions is also on my future list: https://github.com/metaist/cosmofy/issues/94

In theory, anything that can be compiled with libc can be recompiled with cosmopolitan libc, but in practice it takes some work.

**Comment by Spenhouet** [comment]
Is there any bundling solution that has a fast start up when used as CLI tool?

We are still using pyinstaller and it takes for ever for every CLI command.

**Comment by johnthagen** [comment]
> We are still using pyinstaller and it takes for ever for every CLI command.

I've found the penalty is much worse on Windows compared to Linux/macOS, because Windows Defender will scan every file that is unzipped from the PyInstaller exe, and generally has slower file system performance for writings lots of little files.

If I recall, if you use One Folder instead of One File, start up time can be faster: https://pyinstaller.org/en/stable/operating-mode.html#bundling-to-one-folder

since everything doesn't have to be unzipped into a temporary folder each time, but I find [One File](https://pyinstaller.org/en/stable/operating-mode.html#bundling-to-one-file) to be much more convenient for distribution.

All in all, I still use PyInstaller since it's still the best overall experience, but start up time is certainly a downside.

[Python 3.15 lazy imports](https://docs.python.org/3.15/whatsnew/3.15.html#whatsnew315-pep810) might help a _little_ bit in general here for Python CLIs (but doesn't solve the PyInstaller/Windows Defender tax).

**Comment by Spenhouet** [comment]
These solution create a true platform specific executable. We don't necessarily need that.
Yes, we need an easy way to distribute and to automatically provide updates.

I see that the more technical solution (from a user perspective) is to host a platform specific script which is downloaded and run with a one-liner in a terminal and installs everything. This can include a full python installation, install our application as python package, setup auto start (e.g. daemon), setup an auto-update routine regularly pulling for updates, manage python version via uv, register in PATH, ....

But why can't we have a nice installer / UI wrapper for that. Sounds like something totally generic which could work for any python package. This way we wouldn't even need to prebuild everything system compatible but "simply" would install on the respective system. Does that exist already? uv could work nicely under the hood.

**Comment by ceejatec** [comment]
@Spenhouet @johnthagen 

> This way we wouldn't even need to prebuild everything system compatible but "simply" would install on the respective system. Does that exist already? uv could work nicely under the hood.

I wrote wrappers for a tool called "cbdep" I developed at my previous job which is basically this: https://github.com/couchbasebuild/cbdep/tree/master/wrapper

It checks if `uv` is already installed in the usual `$HOME/.local/bin` location, and installs it if not using the Astral-provided script. Then it checks to see if `cbdep` itself is installed AND less than two days old, and if not, uses `uv` to install it. Finally it executes the real `cbdep`, passing all the command-line arguments. In this way, so long as this wrapper is on your `$PATH`, it'll both Just Work(tm) and keep the tool updated. It uses `--python=only-managed` to avoid several [unfortunate interactions](https://github.com/astral-sh/uv/issues/10282) I discovered when `uv` tried to use a system-installed Python - those have since been fixed, but it still made sense to me to avoid the potential for system-dependent trouble.

It's implemented as a simple shell script for Linux and Mac, and as a Go program for Windows. So it can work on any platform, with the only system requirements being bash or Powershell. And the nice thing is that after initial install, it runs at basically native speed since it's only executing locally-stored commands, with no self-extracting archive stuff to set off virus checkers.

It'd be pretty trivial to productize this further, either by making the wrappers auto-install a Python package named `$ARGV[0]`, or else by some very simple code-generation, rather than hard-coding `cbdep`.

Related, I did also raise a [suggestion ](https://github.com/astral-sh/uv/issues/10465) a while ago that it'd be really cool if the `uv` binary itself could work this way by behaving differently when `$ARGV[0]` was different than `uv` or `uvx`. I'd still love to see that, although it has a few limitations as I outlined in the feature request.

**Comment by manatlan** [comment]
Perhaps I'm out of context ...

but a simple : 

    uv run --with pyinstaller pyinstaller --onefile --noconsole my_app.py

does the job asked in the 1st post (thanks to the magic uv) ... no ?

**Comment by stdedos** [comment]
_PS: https://github.com/konchunas/pyrs is archived at Jul 12, 2024; continuation "shown" is https://github.com/py2many/py2many_

**Comment by leandrocm86** [comment]
> Perhaps I'm out of context ...
> 
> but a simple :
> 
> ```
> uv run --with pyinstaller pyinstaller --onefile --noconsole my_app.py
> ```
> 
> does the job asked in the 1st post (thanks to the magic uv) ... no ?

I don't think so because that way you're running uv for pyinstaller, not the script or project, which means the script (or project) dependencies will not be installed by uv. Notice the last argument (my_app.py) would be an argument for pyinstaller, not for uv.

**Comment by Spenhouet** [comment]
By now we also have https://uvx.sh/ available

**Comment by thornewolf** [comment]
> By now we also have https://uvx.sh/ available 

i cannot find a reference to this website from an astral.sh domain. im concerned that this is risky to have linked from this issue if the url is indeed not controlled by astral.

can we get either a confirmation or a deletion so as to prevent people getting tricked by a bad actor on the future?

**Comment by Spenhouet** [comment]
> can we get either a confirmation or a deletion so as to prevent people getting tricked by a bad actor on the future?

<a href="https://x.com/charliermarsh/status/2014417498817348060"><img width="1180" height="1330" alt="Image" src="https://github.com/user-attachments/assets/2dad0532-c781-446d-ba30-fd0daa449746" /></a>

**Comment by stdedos** [comment]
_Someone would say that the screenshot of the same person advocating for a "random" URL is not necessarily proof_

@charliermarsh - could you drop in? 🙏 

(Apologies in advance for the ping 🙏)

**Comment by Spenhouet** [comment]
You can click the image. It links to the original post on Twitter. uvx was also promoted by @tiangolo (where I first saw it). But sure, better if @charliermarsh confirms himself.

**Comment by Dexsys** [comment]
**uv run --active --with pyinstaller pyinstaller --onefile --noconsole my_app.py**

funciona perfectamente, cree un build.bat donde se podría pasar por ejemplo 

**_build.bat my_app_**
```
@echo off
@rem Advertencia, no pasar el **.py** , solo el nombre
uv run --with pyinstaller pyinstaller --onefile --noconsole %1.py >>null
@echo compilación terminada, archivo %1.exe creado.
```
de aquí en adelante se puede mejorar y pasar parámetros... pero muchas gracias por la idea, justo necesitaba esto.

**Comment by kfet** [comment]
> > By now we also have https://uvx.sh/ available
> 
> i cannot find a reference to this website from an astral.sh domain. im concerned that this is risky to have linked from this issue if the url is indeed not controlled by astral.
> 
> can we get either a confirmation or a deletion so as to prevent people getting tricked by a bad actor on the future?

Adding to that - author and domain owner credibility aside (although important) - this is yet another supply chain attack vector, a security nightmare just waiting to happen. Would not recommend using this or a similar approach for anything of substance.

`uv` itself takes great care to make sure what it installs locally can be trusted, and putting a random script in front of it, hosted on the net, with zero ways of making sure it has not been tampered with... is just bonkers.

**Comment by charliermarsh** [comment]
For clarity: yes, I built that as an experiment. We run it and own the domain. It's a Cloudflare worker that emits a shell script to install uv (if not installed), then use uv to install the package, similar to what Mistral did for [vibe](https://raw.githubusercontent.com/mistralai/mistral-vibe/refs/heads/main/scripts/install.sh). We haven't advertised it and don't make any guarantees around it.
